### ISSUE TITLE
`settings` override not being applied when `source` is overriden

### ISSUE THREAD

**Issue by srid** [srid/haskell-flake#238]
PR #221 caused a regression in `settings` overrides not being applied when a package has its `source` set.

Repro : https://github.com/johnhampton/haskell-flake-bug-repo 

Before PR 221,

```
❯ nix develop --override-input haskell-flake github:srid/haskell-flake/c301f31092a74a6acde9d98d6431ea0a9952ca0f                                  │
│warning: not writing modified lock file of flake 'git+file:///tmp/haskell-flake-bug-repo':                                                       │
│• Updated input 'haskell-flake':                                                                                                                 │
│    'github:srid/haskell-flake/1e297173b23c5113dd90a2d299d6e0d864af35f1' (2024-02-21)                                                            │
│  → 'github:srid/haskell-flake/c301f31092a74a6acde9d98d6431ea0a9952ca0f' (2024-01-31)                                                            │
│warning: Ignoring setting 'auto-allocate-uids' because experimental feature 'auto-allocate-uids' is not enabled                                  │
│warning: Ignoring setting 'impure-env' because experimental feature 'configurable-impure-env' is not enabled                                     │
│
```

After PR 221,

```
│❯ nix develop --override-input haskell-flake github:srid/haskell-flake/dd0d3b9d51f226be15367bf6c90006bd0f859f06                                  │
│warning: not writing modified lock file of flake 'git+file:///tmp/haskell-flake-bug-repo':                                                       │
│• Updated input 'haskell-flake':                                                                                                                 │
│    'github:srid/haskell-flake/1e297173b23c5113dd90a2d299d6e0d864af35f1' (2024-02-21)                                                            │
│  → 'github:srid/haskell-flake/dd0d3b9d51f226be15367bf6c90006bd0f859f06' (2024-02-05)                                                            │
│warning: Ignoring setting 'auto-allocate-uids' because experimental feature 'auto-allocate-uids' is not enabled                                  │
│warning: Ignoring setting 'impure-env' because experimental feature 'configurable-impure-env' is not enabled                                     │
│error: builder for '/nix/store/r8dzna69mq5a1b4hbqr3p5nalgsz1p00-openapi3-source-3.2.3.drv' failed with exit code 1;                              │
│       last 10 log lines:                                                                                                                        │
│       > Running phase: updateAutotoolsGnuConfigScriptsPhase                                                                                     │
│       > Running phase: configurePhase                                                                                                           │
│       > configureFlags: --verbose --prefix=/nix/store/6n9r37z5ijhy73izd3dlnxn1wcygrf57-openapi3-source-3.2.3 --libdir=$prefix/lib/$compiler --li│
│bsubdir=$abi/$libname --docdir=/share/doc/openapi3-3.2.3 --with-gcc=gcc --package-db=/build/tmp.KLV8APVgDa/package.conf.d --ghc-options=-j12 +RTS│
│ -A64M -RTS --disable-split-objs --enable-library-profiling --profiling-detail=exported-functions --disable-profiling --enable-shared --disable-c│
│overage --enable-static --disable-executable-dynamic --enable-tests --disable-benchmarks --enable-library-vanilla --disable-library-for-ghci --gh│
│c-option=-split-sections --ghc-options=-haddock --extra-lib-dirs=/nix/store/wyw0d6r22mzl4g7wk1rkif9fsks00qbv-ncurses-6.4/lib --extra-lib-dirs=/ni│
│x/store/zabxhfpsgkb9c4sb7fy50pn1l1kczzv2-libffi-3.4.4/lib --extra-lib-dirs=/nix/store/wanbfij8izz6kjxrqj21gh2fdcb93wg7-gmp-with-cxx-6.3.0/lib --e│
│xtra-lib-dirs=/nix/store/85ybhny02pn4nskrcvvdl1yjpb1zll31-ghc-9.4.8/lib                                                                          │
│       > Using Parsec parser                                                                                                                     │
│       > Configuring openapi3-3.2.3...                                                                                                           │
│       > CallStack (from HasCallStack):                                                                                                          │
│       >   withMetadata, called at libraries/Cabal/Cabal/src/Distribution/Simple/Utils.hs:370:14 in Cabal-3.8.1.0:Distribution.Simple.Utils      │
│       > Error: Setup: Encountered missing or private dependencies:                                                                              │
│       > hspec >=2.5.5 && <2.11                                                                                                                  │
│       >                                                                                                                                         │
│       For full logs, run 'nix log /nix/store/r8dzna69mq5a1b4hbqr3p5nalgsz1p00-openapi3-source-3.2.3.drv'.                                       │
│error: 1 dependencies of derivation '/nix/store/dkjirap2qg3a27xwzj2jrcclfr68y5rg-openapi3-3.2.3.drv' failed to build                             │
│error: 1 dependencies of derivation '/nix/store/jdxgpvdf1dqx3m88fkl4absx4yisw26g-ghc-9.4.8-with-packages.drv' failed to build                    │
│error: 1 dependencies of derivation '/nix/store/wjdv9lch9pphrwm1njmvq6b4i5gjq26f-ghc-shell-for-repo-0.1.0.0-0-env.drv' failed to build
```

cc @johnhampton @Thrithralas 

We need to fix this along with a test case.

**Comment by srid** [comment]
Note: I consider this a serious regression that ought to be fixed in time. But first, I want to improve our testing infra (#241 and others).

**Comment by srid** [comment]
@johnhampton @Thrithralas Can you try out the fix in #252 ?
